buildkite/package_uploads.go

```diff
@@ -0,0 +1,216 @@
+// We do this normalization to ensure that we can accurately calculate the Content-Length of the request body, which is
+// required by S3. We write to disk (instead of buffering in memory) to avoid memory exhaustion for large files.
+func normalizeToFile(r io.Reader, filename string) (*os.File, error) {
+	basename := filepath.Base(filename)
+	f, err := os.CreateTemp("", basename)
+	if err != nil {
+		return nil, fmt.Errorf("creating temporary file: %v", err)
+	}
+
+	_, err = io.Copy(f, r)
+	if err != nil {
+		return nil, fmt.Errorf("writing to temporary file: %v", err)
+	}
+
+	err = f.Close()
+	if err != nil {
+		return nil, fmt.Errorf("closing temporary file: %v", err)
+	}
+
+	// Rename the temporary file to the desired filename, which is important for Buildkite Package indexing
+	newFileName := filepath.Join(filepath.Dir(f.Name()), basename)
+	err = os.Rename(f.Name(), newFileName)
+	if err != nil {
+		return nil, fmt.Errorf("renaming temporary file: %v", err)
+	}
+
+	f, err = os.Open(newFileName)
+	if err != nil {
+		return nil, fmt.Errorf("opening renamed file: %v", err)
+	}
+
+	return f, nil
+}
+
+// PackagePresignedUpload represents a presigned upload URL for a Buildkite package, returned by the Buildkite API
+type PackagePresignedUpload struct {
+	OrganizationSlug string `json:"-"`
+	RegistrySlug     string `json:"-"`
+
+	URI  string                     `json:"uri"`
+	Form PackagePresignedUploadForm `json:"form"`
+}
+
+type PackagePresignedUploadForm struct {
+	FileInput string            `json:"file_input"`
+	Method    string            `json:"method"`
+	URL       string            `json:"url"`
+	Data      map[string]string `json:"data"`
+}
+
+// RequestPresignedUpload requests a presigned upload URL for a Buildkite package from the buildkite API
+func (ps *PackagesService) RequestPresignedUpload(ctx context.Context, organizationSlug, registrySlug string) (PackagePresignedUpload, *Response, error) {
+	url := fmt.Sprintf("v2/packages/organizations/%s/registries/%s/packages/upload", organizationSlug, registrySlug)
+	req, err := ps.client.NewRequest(ctx, "POST", url, nil)
+	if err != nil {
+		return PackagePresignedUpload{}, nil, fmt.Errorf("creating POST presigned upload request: %v", err)
+	}
+
+	var p PackagePresignedUpload
+	resp, err := ps.client.Do(req, &p)
+	if err != nil {
+		return PackagePresignedUpload{}, resp, fmt.Errorf("executing POST presigned upload request: %v", err)
+	}
+
+	p.OrganizationSlug = organizationSlug
+	p.RegistrySlug = registrySlug
+
+	return p, resp, err
+}
+
+// Perform performs uploads the package file referred to by `file` to the presigned upload URL.
+// It does not create the package in the registry, only uploads the file to S3. The returned string is the URL of the
+// uploaded file in S3, which can then be passed to [Finalize] to create the package in the registry.
+func (ppu PackagePresignedUpload) Perform(ctx context.Context, ps *PackagesService, file *os.File) (string, error) {
```

Comment:
this method signature is a little odd compared to the other ones in this library - the `PackagePresignedUpload` is the receiver, and the `*PackagesService` is an arg. should it be the other way around?

my logic here was that the `PackagePresignedUpload` was the key object, and thus "deserved" to be the receiver, but i'm happy to change this to the more standard `*PackagesService` receiver.

One benefit to this approach is that method names are cleaner
```Go
ppu.Perform(ctx, ps, f)
// vs
ps.PerformPresignedUpload(ctx, ppu, f)
```

Reply:
that said though, most of the "complexity" of this API is elided to the user - 99.999% of people are going to call `ps.Create`, which handles all of this from them

Reply:
For total transparency the first approach with `ppu`/`PackagePresignedUpload` receiver and the simpler method names gets my vote. Its more effectively "knowing" that any method on said receiver belongs to a pre-signed upload; whereas `ps` generally could have a whole lot of methods attached which is common across all the other services!

Reply:
good enough for me!